deltachat-ios/AppDelegate.swift

```diff
@@ -382,7 +388,16 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
             if !self.appIsInForeground() {
                 self.dcContext.stopIo()
             }
-            completionHandler(.newData)
+
+            // to avoid 0xdead10cc exceptions, scheduled jobs need to be done before we get suspended;
+            // we increase the probabilty that this happens by adding completionHandler()
+            // to the end of dispatch queues and waiting one second.
+            DispatchQueue.global(qos: .background).async {
+                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
```

Comment:
in the end this asyncAfter (no matter if scheduled from background or not) might be the essential change. 

[the docs](https://developer.apple.com/documentation/dispatch/dispatchqueue/2300077-global) are confusing me:

> Tasks submitted to the returned queue are scheduled concurrently with respect to one another. 

What I conclude: So even if background task A is submitted before background task B, they will run (and finish) concurrently.

I think now, moving the completion handler to the background task will not fix the issue that other background tasks might still be running.

Reply:
Ok, I'm getting more clarity, so the _submission_ of the execution blocks are non-blocking (async) for the global queues. The _execution_ depends on `DispatchQueue.global(qos: .background).sync / DispatchQueue.global(qos: .background).async`

Reply:
Since we always dispatch queues to be executed asynchronously (`DispatchQueue.global(qos: .background).async`) we have no guarantee that the completion handler will be executed and finish last in the background queue

Reply:
If we would implement our own DispatchQueues, we could make use of the .barrier flag (they have no effect in global queues). That is, in a async queue, all tasks submitted before the execution block marked with the .barrier flag, would finish before that execution block is executed.

Reply:
yip, the docs are confusing. as so often :)

but anyway, on av we decided to just rely on a timeout and see if that actually fixes the issue. next escalation would be to tweak DispatchQueue.global and maybe even adding a dc_suspend()